rule/sexpr/lexer.go

```diff
@@ -64,3 +73,385 @@ func isComment(r rune) bool {
+		// number, but some other form.
+		rn, err := s.readRune()
+
+		if err != nil {
+			se := err.(*ScanError)
+			if se.EOF {
+				// In reality having '-' as the final
+				// symbol in a stream is never useful,
+				// but this is the sort of error we
+				// should catch in the Parser, not the
+				// scanner.
+				return SYMBOL, b.String(), nil
+			}
+			return NUMBER, b.String(), err
+		}
+
+		// We've stored the rune, and we know we'll want to
+		// unread whatever happens, so lets just do that now.
+		s.unreadRune(rn)
+
+		// If the next rune isn't a digit then we're going to
+		// assume this is the minus operator and return '-' as
+		// a symbol instead of a number.  There are still
+		// cases where this wouldn't be valid, but they're all
+		// errors and we'll leave that for the Parser to
+		// handle.
+		if !unicode.IsDigit(rn) {
+			return SYMBOL, b.String(), nil
+		}
+	}
+
+	// OK, let's scan the rest of the number...
+
+	for {
+		rn, err := s.readRune()
+		if err != nil {
+			se := err.(*ScanError)
+			if se.EOF {
+				// EOF is a valid terminator for a number
+				return NUMBER, b.String(), nil
+			}
+			return NUMBER, b.String(), err
+		}
+		if rn == '-' {
+			// As we said before '-' can't appear in the
+			// body of a number, this is an error.
+			return NUMBER, b.String(), s.newScanError("invalid number format (minus can only appear at the beginning of a number)")
+		}
+
+		// Valid number parts are written to the buffer
+		if isNumber(rn) || rn == '.' {
+			b.WriteRune(rn)
+			continue
+		}
+		// we hit a terminating character, end the number here.
+		s.unreadRune(rn)
+		break
+	}
+	return NUMBER, b.String(), nil
+}
+
+// scanBool scans the contiguous characters following the '#' symbol,
+// it they are either 'true', or 'false' a BOOL is returned, otherwise
+// an ScanError will be returned.
+func (s *Scanner) scanBool() (Token, string, error) {
+
+	var b bytes.Buffer
+
+	for {
+		rn, err := s.readRune()
+		if err != nil {
+			se := err.(*ScanError)
+			if se.EOF {
+				// EOF is a valid terminator for a number
+				break
+			}
+			return BOOL, b.String(), err
+		}
+
+		// isSymbol is handy shorthand for "it's not anything else"
+		if !isSymbol(rn) {
+			s.unreadRune(rn)
+			break
+		}
+		b.WriteRune(rn)
+	}
+
+	symbol := b.String()
+	if symbol == "true" || symbol == "false" {
+		return BOOL, symbol, nil
+	}
+	if len(symbol) > 0 {
+		return BOOL, symbol, s.newScanError(fmt.Sprintf("invalid boolean: %s", symbol))
+	}
+	return BOOL, symbol, s.newScanError("invalid boolean")
+}
+
+// scanComment will scan to the end of the current line, consuming any and all chars prior to '\n'.
+func (s *Scanner) scanComment() (Token, string, error) {
+	var b bytes.Buffer
+
+	for {
+		rn, err := s.readRune()
+		if err != nil {
+			se := err.(*ScanError)
+			if se.EOF {
+				// EOF is a valid terminator for a Comment
+				break
+			}
+			return COMMENT, b.String(), err
+		}
+
+		if rn == '\n' {
```

Comment:
Does this also catch `\r\n`?

Reply:
Yes.  Because `\r` is really quite meaningless and it's always followed by a `\n`. we can just say `\r` is a white space character, but it has no impact on the line count.

Reply:
But the returned string will contain `blablabla\r`, is it an issue?

Reply:
cc @tealeg

Reply:
I've made an issue to remove the embedded `\r` from the comment anyway:
https://github.com/heetch/regula/issues/41